frontends/p4/moveDeclarations.h

```diff
@@ -36,17 +36,22 @@ class MoveDeclarations : public Transform {
 
     /// List of lists of declarations to move, one list per
     /// control/parser/action.
-    std::vector<IR::Vector<IR::Declaration> *> toMove;
-    void push() { toMove.push_back(new IR::Vector<IR::Declaration>()); }
+    // FIXME: Do we really need IR::Vector here?
+    std::vector<IR::Vector<IR::Declaration>> toMove;
```

Comment:
Frankly speaking, I do not know. `Vector` is 112 bytes for me (as compared to 24 bytes of just `std::vector`), so it is definitely large compared to plain vector. The only question is how many `Vector`'s we might have for large apps.

Reply:
Hmm -- IR::Vector should be 64 bytes larger on a 64 byte machine.  Most of that (48 bytes) is for the SourceInfo which is larger than it should be.   If the difference is more than that, it implies the C++ compiler is doing something odd (such as using unnecessary redundant pointers in the object for mulitple inheritance).

Reply:
@ChrisDodd 

`sizeof(Node) == 88` for me:
 - 24 bytes of `INode`
 - 48 bytes of `SourceInfo`
 - 2x4 bytes for `id` and `cloneId`
 - And `8` bytes for vtable pointer (essentially from `RTTI::Base`)

`INode` is 24 bytes because it is having 3 base classes each with virtual methods, so need to hold 3 pointers to vtables.

`Vector` is:
  - `VectorBase`: 64 bytes (essentially, `Node` w/o `INode`)
  - `safe_vector`: 24 bytes
  - `INode`: 24 bytes

Reply:
Yes, INode should be 0 bytes for a reasonable C++ implementation -- everything for the virtual base classes can be in Node's vtable.  The fact that it is wasting 24 bytes in every object is annoying.

Reply:
How it will be implemented then? E.g.:

```c++
class IFoo {
 virtual void foo() = 0;
}

class IBar {
 virtual void bar() = 0;
}

class IBaz {
 virtual void baz() = 0;
}

class IBig : public IFoo, public IBar, public IBaz {
};

class Foo : public virtual IBig {
  some fields;
  virtual void aaaa();
  void foo() override;
  void bar() override;
  void baz() override;
}

Foo f;

void callBar(IBar *b) {
  b->bar();
}

void callBar2(Foo *f) {
  f->bar();
}

callBar(&b);
callBar(f);
```

How we'd make both cases work?

Reply:
Each of the bases has to have its own `vptr` if it has any virtual member functions, there is no way around this (as a pointer through the base type has to be valid for each base and has to be able to access vptr and members of the base). The slide below[^1] demonstrates the layout in a particular clang implementation (I'd say GCC on Linux has to do the same).

![image](https://github.com/user-attachments/assets/a582ce10-aa24-4dcb-a8bd-5e0ee3596302)

All the offsets (for casting etc.) are kept in `vptr` and so are the member functions, but there must be multiple `vptr`s. The implementation is actually pretty close to what Bjarne Stroustrup published in his paper on multiple inheritance back before C++ was standardized.

[^1]: I created the slice few years ago for an advanced C++ course at [fi.muni.cz](https://fi.muni.cz).

Reply:
> How we'd make all cases work?

It requires laying out the vtables at link time.  That requires either a smarter linker, or a common virtual base class for IFoo/IBar/IBaz, in which case you can have a linker section for laying out the vtables of all subclasses of that common virtual base class with even a dumb traditional linker.

Reply:
But you can load a class from a `.so` file at runtime and use it in a program, and it can have a combination of bases from the original program and from the `.so`. And you have to know sizes of the structs much earlier than that happens.